tfjs-backend-webgpu/src/conv_backprop_mm_webgpu.ts

```diff
@@ -16,107 +16,168 @@
+      if (xC < 0.0 || xC >= f32(uniforms.outBackprop[2]) || fract(xC) > 0.0) {
+        return ${typeSnippet(innerElementSize)}(0.0);
+      }
+      let coord = vec4<i32>(
+          batch,
+          i32(xR),
+          i32(xC),
+          col % uniforms.outBackprop[3]);
+      return x[getIndexFromCoords4D(coord, uniforms.xShape)/${
+      innerElementSize}];`;
+
+  const sampleA = `if (row < uniforms.dimAOuter && col < uniforms.dimInner) {
+        ${readASnippet}
+      }
+      return ${typeSnippet(innerElementSize)}(0.0);`;
+
+  const userCode = `
+  fn mm_readA(row : i32, colIn : i32, globalId : vec3<u32>) -> ${
+      typeSnippet(innerElementSize)} {
+    let col = colIn * ${innerElementSize};
+    var batch = i32(globalId.z);
+    ${sampleA}
+  }
+
+  fn mm_readB(row : i32, colIn : i32, globalId : vec3<u32>) -> ${
+      typeSnippet(innerElementSize)} {
+    let col = colIn * ${innerElementSize};
+    let coordX = uniforms.filterDims.x - 1 -
+        row / (uniforms.filterDims[1] * uniforms.outBackprop[3]);
+    let coordY = uniforms.filterDims.y - 1 -
+        (row / uniforms.outBackprop[3]) % uniforms.filterDims[1];
+    if (row < uniforms.dimInner && col < uniforms.dimBOuter &&
+        coordX >= 0 && coordY >= 0) {
+      let rowInner = row % uniforms.outBackprop[3];
+      let coord = vec4<i32>(coordX, coordY, col, rowInner);
+      ${getWSnippet(innerElementSize)}
+    }
+    return ${typeSnippet(innerElementSize)}(0.0);
+  }
+
+  fn mm_write(row : i32, colIn : i32, valueInput : ${
+      typeSnippet(innerElementSize)}, globalId : vec3<u32>) {
+    let col = colIn * ${innerElementSize};
+    if (row < uniforms.dimAOuter && col < uniforms.dimBOuter) {
+      var batch = i32(globalId.z);
+      var value = valueInput;
+      let outCoord = vec4<i32>(
+          batch,
+          row / uniforms.outShape[2],
+          row % uniforms.outShape[2],
+          col);
+      result[getIndexFromCoords4D(outCoord, uniforms.outShape)/${
+      innerElementSize}] = value;
+    }
+  }`;
+  return userCode;
+}
+
 export class Conv2DDerInputMMProgram implements WebGPUProgram {
   outputShape: number[];
   shaderKey: string;
   dispatchLayout: {x: number[], y: number[], z: number[]};
   dispatch: [number, number, number];
   variableNames = ['x', 'W'];
+  variableTypes: string[];
   uniforms =
       'filterDims : vec2<i32>, pads : vec2<i32>, stride : vec2<i32>, outBackprop : vec4<i32>, dimAOuter : i32, dimBOuter : i32, dimInner : i32,';
   workGroupSize: [number, number, number];
   elementsPerThread: [number, number, number];
+  isChannelsLast: boolean;
+  tileAOuter: number;
+  tileBOuter: number;
+  tileInner: number;
+  innerElementSize: number;
+  isVec4?: boolean;
 
   constructor(convInfo: backend_util.Conv2DInfo) {
     this.outputShape = convInfo.inShape;
 
     util.assert(
         convInfo.dataFormat === 'channelsLast',
         () => 'TODO: NCHW is unimplemented');
+    this.isChannelsLast = convInfo.dataFormat === 'channelsLast';
```

Comment:
nit: Currently `NCHW` is unimplemented. So `this.isChannelsLast` can be ignored in this PR.

Reply:
done